exporter/trace/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,290 @@
+// If no custom hook is set, errors are logged.
+func WithOnError(onError func(err error)) func(o *options) {
+	return func(o *options) {
+		o.OnError = onError
+	}
+}
+
+// WithTraceClientOptions sets additionial client options for tracing.
+func WithTraceClientOptions(opts []option.ClientOption) func(o *options) {
+	return func(o *options) {
+		o.TraceClientOptions = opts
+	}
+}
+
+// WithContext sets the context that trace exporter and metric exporter
+// relies on.
+func WithContext(ctx context.Context) func(o *options) {
+	return func(o *options) {
+		o.Context = ctx
+	}
+}
+
+func (o *options) handleError(err error) {
+	if o.OnError != nil {
+		o.OnError(err)
+		return
+	}
+	log.Printf("Failed to export to Stackdriver: %v", err)
+}
+
+// defaultTimeout is used as default when timeout is not set in newContextWithTimout.
+const defaultTimeout = 5 * time.Second
+
+// Exporter is a trace exporter that uploads data to Stackdriver.
+//
+// TODO(yoshifumi): add a metrics exporter once the spec definition
+// process and the sampler implementation are done.
+type Exporter struct {
+	traceExporter *traceExporter
+	spanProcessor trace.SpanProcessor
+}
+
+// NewExporter creates a new Exporter thats implements trace.Exporter.
+//
+// TODO(yoshifumi): add a metrics exporter one the spec definition
+// process and the sampler implementation are done.
+func NewExporter(opts ...Option) (*Exporter, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+	if o.ProjectID == "" {
+		ctx := o.Context
+		if ctx == nil {
+			ctx = context.Background()
+		}
+		creds, err := google.FindDefaultCredentials(ctx, traceapi.DefaultAuthScopes()...)
+		if err != nil {
+			return nil, fmt.Errorf("Stackdriver: %v", err)
+		}
+		if creds.ProjectID == "" {
+			return nil, errors.New("Stackdriver: no project found with application default credentials")
+		}
+		o.ProjectID = creds.ProjectID
+	}
+	te, err := newTraceExporter(&o)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Exporter{
+		traceExporter: te,
+	}, nil
+}
+
+func newContextWithTimeout(ctx context.Context, timeout time.Duration) (context.Context, func()) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	if timeout <= 0 {
+		timeout = defaultTimeout
+	}
+	return context.WithTimeout(ctx, timeout)
+}
+
+// RegisterBatchSpanProcessor registers e as BatchSpanProcessor.
+func (e *Exporter) RegisterBatchSpanProcessor() error {
+	opts := []trace.BatchSpanProcessorOption{}
+
+	if e.traceExporter.o.BatchDelayThreshold > 0 {
+		opts = append(opts, trace.WithScheduleDelayMillis(e.traceExporter.o.BatchDelayThreshold))
+	}
+	if e.traceExporter.o.MaxQueueSize > 0 {
+		opts = append(opts, trace.WithMaxQueueSize(e.traceExporter.o.MaxQueueSize))
+	}
+	if e.traceExporter.o.MaxExportBatchSize > 0 {
+		opts = append(opts, trace.WithMaxExportBatchSize(e.traceExporter.o.MaxExportBatchSize))
+	}
+
+	var err error
+	e.spanProcessor, err = trace.NewBatchSpanProcessor(e, opts...)
```

Comment:
This should be protected using a lock and should be allowed only once.

Reply:
Embedded sync.Once in Exporter to lock and do the initialization only once.